sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -739,15 +739,24 @@ object HashJoin extends CastSupport with SQLConfHelper {
     // jump over keys that have a higher index value than the required key
     if (keys.size == 1) {
       assert(index == 0)
-      cast(BoundReference(0, LongType, nullable = false), keys(index).dataType)
+      Cast(
+        child = BoundReference(0, LongType, nullable = false),
+        dataType = keys(index).dataType,
+        timeZoneId = Option(conf.sessionLocalTimeZone),
+        ansiEnabled = false)
```

Comment:
so we do expect to get null value if overflow happens?

cc @c21

Reply:
The default cast returns the lower bits when an overflow happens.
E.g.
```
> select cast(1025L as byte);
1
```

Reply:
Sorry just for my understanding (without running the unit test by myself): why we expect overflow happens here? Shouldn't `rewriteKeyExpr()` above already guarantees that we only rewrite keys to long when all keys fit into a long type?

Reply:
@c21 this PR is to avoid exceptions on overflow.
If you run HashedRelationSuite with ANSI mode on, there will be an overflow error

Reply:
ah I see, so this algorithm expects the equivalent of java code `byte b = (byte) long_value`, which is the same as the non-ansi Cast behavior.